python/ray/serve/config.py

```diff
@@ -263,84 +265,27 @@ def __init__(
-            "max_pending_calls",
-            "scheduling_strategy",
-        }
-
-        for option in disallowed_ray_actor_options:
-            if option in self.ray_actor_options:
-                raise ValueError(
-                    f"Specifying {option} in ray_actor_options is not allowed."
-                )
-
-        # TODO(suquark): reuse options validation of remote function/actor.
-        # Ray defaults to zero CPUs for placement, we default to one here.
-        if self.ray_actor_options.get("num_cpus") is None:
-            self.ray_actor_options["num_cpus"] = 1
-        num_cpus = self.ray_actor_options["num_cpus"]
-        if not isinstance(num_cpus, (int, float)):
-            raise TypeError("num_cpus in ray_actor_options must be an int or a float.")
-        elif num_cpus < 0:
-            raise ValueError("num_cpus in ray_actor_options must be >= 0.")
-        self.resource_dict["CPU"] = num_cpus
-
-        if self.ray_actor_options.get("num_gpus") is None:
-            self.ray_actor_options["num_gpus"] = 0
-        num_gpus = self.ray_actor_options["num_gpus"]
-        if not isinstance(num_gpus, (int, float)):
-            raise TypeError("num_gpus in ray_actor_options must be an int or a float.")
-        elif num_gpus < 0:
-            raise ValueError("num_gpus in ray_actor_options must be >= 0.")
-        self.resource_dict["GPU"] = num_gpus
-
-        # Serve deployments use Ray's default for actor memory.
-        self.ray_actor_options.setdefault("memory", None)
-        memory = self.ray_actor_options["memory"]
-        if memory is not None and not isinstance(memory, (int, float)):
-            raise TypeError(
-                "memory in ray_actor_options must be an int, a float, or None."
-            )
-        elif memory is not None and memory <= 0:
-            raise ValueError("memory in ray_actor_options must be > 0.")
-        self.resource_dict["memory"] = memory
-
-        object_store_memory = self.ray_actor_options.get("object_store_memory")
-        if not isinstance(object_store_memory, (int, float, type(None))):
-            raise TypeError(
-                "object_store_memory in ray_actor_options must be an int, float "
-                "or None."
-            )
-        elif object_store_memory is not None and object_store_memory < 0:
-            raise ValueError("object_store_memory in ray_actor_options must be >= 0.")
-        self.resource_dict["object_store_memory"] = object_store_memory
-
-        if self.ray_actor_options.get("resources") is None:
-            self.ray_actor_options["resources"] = {}
-        custom_resources = self.ray_actor_options["resources"]
-        if not isinstance(custom_resources, dict):
-            raise TypeError("resources in ray_actor_options must be a dictionary.")
-        self.resource_dict.update(custom_resources)
+        self.resource_dict = resources_from_ray_options(self.ray_actor_options)
+        self.resource_dict.setdefault("CPU", 1)
```

Comment:
We also need to set the default `"num_cpus"` in `ray_actor_options` to 1. Make sure not to use `setdefault` for that because even if `"num_cpus"` is explicitly set to `None`, it should still default to 1.

Reply:
In `resources_from_ray_options`, `CPU` will never be `None`. The `CPU` key will be removed if the input is `None`.

https://github.com/ray-project/ray/blob/e402fc0eaa07d3dd396730661bb799e5aab9a03a/python/ray/_private/utils.py#L307-L349

Also Ray does not set `num_cpus=None` automatically if it is not specified by users, after recent updates.

This design was to differentiate "default CPU value" and "CPU value not set". Not differentiating "default value" and "value not set" is a very bad idea, because 1. you cannot catch the bug that you forget to set a default value 2. it confuses the dependent code who want to change the options, because it does not know whether the value was set initially. For example: https://github.com/ray-project/ray/pull/24131/files

Reply:
Ah, I see. For more context, [Serve contains JSON schemas](https://github.com/ray-project/ray/blob/6151b75d9df851f98a0f3bfdbd98201a81a25b30/python/ray/serve/schema.py#L8-L74) that use `None` as default for the `ray_actor_options`. This is why the `ReplicaConfig`'s `_validate()` function converts `None` to default values.

I think it makes more sense to simply omit the defaults in the schema. Could you remove the defaults from the [RayActorOptionsSchema](https://github.com/ray-project/ray/blob/6151b75d9df851f98a0f3bfdbd98201a81a25b30/python/ray/serve/schema.py#L8-L74)?

Also, we should still explicitly set `num_cpus`'s default to 1. That's a Serve default that I believe differs from Ray's default. That means even if `ray_actor_options is None`, we should still insert `"num_cpus": 1` into the dictionary. And if it's not `None`, but it doesn't have a `"num_cpus"` field, we should add `"num_cpus": 1` to the dictionary. The `ray_actor_options` dictionary is what ultimately [gets passed into each replica actor's `.options()` call](https://github.com/ray-project/ray/blob/859f8cb0ec5fcea562862cd66a80d248af96e3dc/python/ray/serve/deployment_state.py#L329). The `resource_dict` is used only to inform the user about their [resource usage](https://github.com/ray-project/ray/blob/859f8cb0ec5fcea562862cd66a80d248af96e3dc/python/ray/serve/deployment_state.py#L751-L770).

Reply:
Thanks for pointing out the `ray_actor_options` issue! For the schema, I will keep the schema as it is, because although we does not use `None` as default values, `None` is still a valid input value for backwards compatibility. Removing it from the default values would cause failures when processing `None` inputs.